client/services/code_preprocessors/PythonCodePreprocessorService.js

```diff
@@ -152,117 +160,181 @@ tie.factory('PythonCodePreprocessorService', [
+            VARNAME_TASK_CORRECTNESS_TEST_RESULTS + ' = []');
+
+        for (var j = 0; j < oneTaskCorrectnessTests.length; j++) {
+          var testInputCode = (
+            inputFunctionName ?
+            inputFunctionName + '(' + VARNAME_ALL_TASKS_TEST_INPUTS + '[' + i + '][' + j + '])' :
+            VARNAME_ALL_TASKS_TEST_INPUTS + '[' + i + '][' + j + ']'
+          );
+          var testRunCode = (
+            'System.runTest(' + qualifiedMainFunctionName + ', '
+            + testInputCode + ')'
+          );
+          var testOutputCode = (
+            outputFunctionName ?
+            outputFunctionName + '(' + testRunCode + ')' : testRunCode);
+
+          testCode += (
+            '\n' +
+            VARNAME_TASK_CORRECTNESS_TEST_RESULTS +
+            '.append(' + testOutputCode + ')');
+        }
 
         testCode += (
           '\n' +
-          VARNAME_CORRECTNESS_TEST_RESULTS + '.append(' + testOutputCode + ')');
+          VARNAME_CORRECTNESS_TEST_RESULTS +
+          '.append(' + VARNAME_TASK_CORRECTNESS_TEST_RESULTS + ')');
       };
 
       return testCode;
     };
 
     var _generateBuggyOutputTestCode = function(
-        buggyOutputTests, inputFunctionName, outputFunctionName) {
+        allTasksBuggyOutputTests,
+        allTasksInputFunctionNames, allTasksOutputFunctionNames) {
       // NOTE: This must be run after the correctness tests, since it assumes
       // that the test inputs and correctness test results already exist.
       // TODO(sll): Cache the results of running the buggy code, so that they
       // don't have to be recomputed for every run.
-      var testInputCode = (
-        inputFunctionName ?
-        inputFunctionName + '(test_input)' :
-        'test_input'
-      );
-      var testRunCode = 'System.runTest(func, ' + testInputCode + ')';
-      var testOutputCode = (
-        outputFunctionName ?
-        outputFunctionName + '(' + testRunCode + ')' : testRunCode);
-
-      var fullTestCode = [
-        'def matches_buggy_function(func):',
-        '    buggy_results = []',
-        '    for test_input in test_inputs:',
-        '        buggy_results.append(' + testOutputCode + ')',
-        '    return buggy_results == ' + VARNAME_CORRECTNESS_TEST_RESULTS,
-        '',
-        VARNAME_BUGGY_OUTPUT_TEST_RESULTS + ' = []',
-        ''
-      ].join('\n');
-      buggyOutputTests.forEach(function(buggyOutputTest) {
-        var qualifiedBuggyFunctionName = (
-          buggyOutputTest.getBuggyFunctionName());
-
-        fullTestCode += (
-          VARNAME_BUGGY_OUTPUT_TEST_RESULTS +
-          '.append(matches_buggy_function(' +
-          qualifiedBuggyFunctionName + '))\n');
-      });
+      var matchesBuggyFunctionSet = new Set();
+
+      for (var i = 0; i < allTasksInputFunctionNames.length; i++) {
+        var inputFunctionName = allTasksInputFunctionNames[i];
+        var outputFunctionName = allTasksOutputFunctionNames[i];
+
+          var testInputCode = (
```

Comment:
The indentation here and below seems wrong -- it should be on the same level as the previous line of code.